add a blog section on `code.groundlight.ai` and a blog post on framegrab (#138)
groundlight/python-sdk: docs/blog/2023-12-06-framegrab.md

---
title: Announcing Groundlight's FrameGrab Library

> description: Today, we are happy to announce FrameGrab 0.4.3. 

Reviewer: What are the v0.4.3 specific features?

tags: [groundlight-extensions, framegrab]
image: https://i.imgur.com/mErPwqL.png
hide_table_of_contents: false
---

<!-- truncate -->

At Groundlight, we continue to build infrastructure that allows our customers to easily use computer 
vision without a pre-existing dataset for industrial inspection, retail analytics, mobile robotics, and 
much more. We've built many features towards the goal of declarative computer vision, and today we are excited to 
announce FrameGrab, an extremely easy-to-use Python library designed to make it easy to grab frames from
cameras or streams. 

FrameGrab supports generic USB cameras, RTSP streams, Basler USB cameras, Basler GigE cameras, and 
Intel RealSense depth cameras. 

## Grabbing Camera Frames

Frame grabber objects are configured through YAML. The configuration combines the camera type, camera ID, and the camera
options. The YAML config contains many configurable features, but only `input_type` is required. Valid choices for 
`input_type` include 

* generic_usb
* rtsp
* realsense
* basler 

Here is an example of how to use the generic USB configuration 

> ```python 

Author: I went with the `python notest` option.

from framegrab import FrameGrabber 

config = """
name: Front Door Camera
input_type: generic_usb
id:
  serial_number: 23432570
options:
    resolution:
        height: 1080
        width: 1920
    zoom:
        digital: 1.5
"""

grabber = FrameGrabber.create_grabber_yaml(config)
frame = grabber.grab()

# Do real work with the frame 

# Finally release the grabber object 

> grabber.release()

Author: Good suggestion!

for grabber in grabbers.values():
    print(grabber.config)

    # Do real work 

    # Release the frame grabber object 
    grabber.release()

```

## Using FrameGrab for Motion Detection 

With this release, we also continue to support [motion detection](https://en.wikipedia.org/wiki/Motion_detection) via the frame differencing
algorithm. This is an extremely fast algorithm for easily detecting motion in a sequence of frames. 

To use motion detection, it suffices to initialize the MotionDetector instance with the desired percentage of pixels 
needed to change in an image for it to be flagged for motion and the minimum brightness change for each pixel for it 
to be considered changed. Here is a comprehensive example. 

```python 
from framegrab import FrameGrabber, MotionDetector

config = {
    'input_type': 'webcam',
}
grabber = FrameGrabber.create_grabber(config)
motion_detector = MotionDetector(pct_threshold=motion_threshold, val_threshold=60)

while True:
    frame = grabber.grab()
    if frame is None:
        print("No frame captured!")
        continue

    if motion_detector.motion_detected(frame):
        print("Motion detected!")

```

## Conclusion 

> This release of FrameGrab comes with several awesome features that are very easy to use. We now support 

Reviewer: Trying to reduce the repetition of the word 'awesome'
```suggestion
This release of FrameGrab adds some great, easy to use features. We now support 
```
